File: leetcode/asteroid-collision.py
from typing import List


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for a in asteroids:
            add_to_stack = True
            while stack and a < 0 and stack[-1] > 0:
                if abs(a) == stack[-1]:
                    add_to_stack = False
                    stack.pop()
                    break
                elif abs(a) > stack[-1]:
                    stack.pop()
                else:
                    add_to_stack = False
                    break

            if add_to_stack:
                stack.append(a)

        return stack

File: leetcode/asteroid-collision.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<int> asteroidCollision(vector<int>& asteroids)
    {
        vector<int> st;

        for (auto a : asteroids) {
            auto insert = true;
            while (!st.empty() && st.back() > 0 && a < 0) {
                if (st.back() < abs(a)) {
                    st.pop_back();
                } else if (st.back() == abs(a)) {
                    st.pop_back();
                    insert = false;
                    break;
                } else {
                    insert = false;
                    break;
                }
            }

            if (insert) {
                st.push_back(a);
            }
        }

        return st;
    }
};
